koku/cost_models/serializers.py

```diff
@@ -54,6 +55,44 @@ class MarkupSerializer(serializers.Serializer):
     unit = serializers.ChoiceField(choices=MARKUP_CHOICES, required=False)
 
 
+class DistributionSerializer(BaseSerializer):
+    """Serializer for distribution options"""
+
+    DISTRIBUTION_OPTIONS = ("distribute_type", "worker_cost", "platform_cost")
+
+    distribute_type = serializers.ChoiceField(choices=metric_constants.DISTRIBUTION_CHOICES, required=False)
+    platform_cost = serializers.BooleanField(required=False)
+    worker_cost = serializers.BooleanField(required=False)
+
+    def validate_distribute_type(self, distribute_type):
+        """Run validation for distribute type."""
+        distrib_choice_list = [choice[0] for choice in metric_constants.DISTRIBUTION_CHOICES]
+        if distribute_type not in distrib_choice_list:
+            error_msg = f"{distribute_type} is an invaild distribution type"
+            raise serializers.ValidationError(error_msg)
+        return distribute_type
+
+    def validate(self, data):
+        """Run validation for distribution options."""
+
+        if not any(ele in data.keys() for ele in self.DISTRIBUTION_OPTIONS):
+            data = data.get("distribution_info", {})
+        if data.keys() and not (sorted(self.DISTRIBUTION_OPTIONS) == sorted(data.keys())):
+            distribution_info_str = ", ".join(str(option) for option in self.DISTRIBUTION_OPTIONS)
+            error_msg = f"Missing distribution information: one of {distribution_info_str}"
+            raise serializers.ValidationError(error_msg)
+        else:
+            _distribute_type = data.get("distribute_type", metric_constants.CPU_DISTRIBUTION)
+            valid_distribute_type = self.validate_distribute_type(_distribute_type)
+            if data == {}:
+                data["distribute_type"] = valid_distribute_type
+                data["platform_cost"] = True
+                data["worker_cost"] = True
+            else:
+                data["distribute_type"] = valid_distribute_type
+        return data
```

Comment:
Almost none of this is required. These serializers have validation built in. So for the `distribute_type`, when the serializer receives the data, it automatically validates the fields and they can only be one of the `metric_constants.DISTRIBUTION_CHOICES`. So that method can be removed entirely. As for the validation method, the only thing we need to do is ensure we have all 3 fields if the user passes in the json. This validate method can be replaced with:

```
DISTRIBUTION_OPTIONS = {"distribute_type", "worker_cost", "platform_cost"} <<<< this is a set so we can do set math

def validate(self, data):
    diff = self.DISTRIBUTION_OPTIONS.difference(data) # this gets the difference between the options and the data
    if diff == self.DISTRIBUTION_OPTIONS:  # here, if the diff is the same as the options, it means we've received an empty dict, so we just need to populate the default data and return it. It might be better to create the default as a constant somewhere and return the constant instead.
        return {
            "distribute_type": metric_constants.CPU_DISTRIBUTION,
            "platform_cost": True
            "worker_cost": True
        }
    if diff:  # if there is a diff that does _not_ match the options, that means things are missing, which is a validation error.
        distribution_info_str = ", ".join(diff)  # `join` takes an iterable and all the entries of self.DISTRIBUTION_OPTIONS are defined as strings, so there is no need to use `str()` here, or iterate over the entries
            error_msg = f"Missing distribution information: {distribution_info_str}"  # instead of giving a general, "hey one of these is missing", this error will tell the user which field is actually missing
            raise serializers.ValidationError(error_msg)
    return data  # if we made it this far, we just need to return the data we have because the individual validators have already validated the individual entries of the json
```

Reply:
I like this approach much better. Using `set` simplifies things a lot here, and it makes is much simpler to provide a useful error message.

Reply:
I agree with this as well, thanks @maskarb 👍🏿 
I've updated that validate method